Richer C/C++ syntax highlighting.

diff --git a/syntax/c.nanorc b/syntax/c.nanorc
index 905587c..9c759d0 100644
--- a/syntax/c.nanorc
+++ b/syntax/c.nanorc
@@ -5,46 +5,48 @@ header "-\*-.*\<C(\+\+)?((;|[[:blank:]]).*)?-\*-"
 magic "^(C|C\+\+) (source|program)"
 comment "//"
 
-# Constants.
-color brightred "\<[[:upper:]_][[:upper:][:digit:]_]*\>"
-# Labels.
-color brightmagenta "^[[:blank:]]*[[:alpha:]_][[:alnum:]_]*:[[:blank:]]*$"
-color normal ":[[:blank:]]*$"
-
-# Types and related keywords.
-color green "\<(bool|char|double|enum|float|int|short|(un)?signed|struct|union|void|((u_?)?int(8|16|32|64)|[[:lower:]][[:lower:]_]*)_t)\>"
-color green "\<(align(as|of)|auto|const(eval|expr|init)?|extern|inline|register|restrict|sizeof|static(_assert)?|thread_local|typedef|typeof(_unqual)?|volatile)\>"
-color green "\<(_(Align(as|of)|Atomic|BitInt|Bool|Complex|Decimal(32|64|128)|Generic|Imaginary|Noreturn|Static_assert|Thread_local))\>"
-color green "\<(class|explicit|final|friend|mutable|namespace|override|private|protected|public|template|this|typename|using|virtual)\>"
-color green "\<(char(8|16|32)_t|wchar_t|(const|dynamic|reinterpret|static)_cast|decltype|export|import|module|noexcept|requires|typeid)\>"
-
-# Flow control.
-color brightyellow "\<(if|else|for|while|do|switch|case|default)\>"
-color brightyellow "\<(try|throw|catch|operator|new|delete)\>"
-color magenta "\<(break|continue|goto|(co_)?return)\>"
-
-# Single-quoted stuff (characters, backslash escapes, hex and octal byte codes).
-color brightmagenta "'([^'\]|\\(["'\abfnrtv]|x[[:xdigit:]]{1,2}|[0-3]?[0-7]{1,2}))'"
-
-# Hexadecimal, binary, and boolean values.
-color pink "\<(0[Xx][[:xdigit:]]['[:xdigit:]]*|0[Bb][01]['01]*|true|false|nullptr)\>"
-
-# GCC builtins.
-color cyan "__attribute__[[:blank:]]*\(\([^)]*\)\)|__(aligned|asm|builtin|hidden|inline|packed|restrict|section|typeof|weak)__"
-
-# Strings and names of included files.
-color brightyellow ""([^"]|\\")*"|#[[:blank:]]*include[[:blank:]]*<[^>]+>"
+# Macros.
+color brightred "\<[A-Z_][0-9A-Z_]+\>"
+
+# Keywords.
+color green "\<(float|double|bool|char|int|short|long|sizeof|typeid|enum|void|auto|static|register|const|volatile|struct|union|typedef|extern|(un)?signed|inline)\>"
+color green "\<(class|namespace|template|public|protected|private|typename|friend|virtual|noexcept|using|mutable|explicit|export|(wchar|char(8|16|32))_t)\>"
+color green "\<(align(as|of)|decltype|constexpr|consteval|thread_local|restrict|_Complex|_Bool|_Imaginary|_Atomic|_Thread_local|_Noreturn|_Align(as|of))\>"
+color green "\<(operator|new|delete|(static|dynamic|const|reinterpret)_cast|static_assert|_Static_assert)\>"
+color brightyellow "\<(for|if|while|do|else|case|default|switch|try|catch)\>"
+color brightmagenta "\<(goto|continue|break|return|throw)\>"
+color yellow "\<(true|false|this|nullptr|__func__)\>"
+
+# Special identifiers.
+color red "\<([[:lower:]][[:lower:]_]*|s?size|ptrdiff|u?int((_(fast|least))?(8|16|32|64)|ptr|max))_t\>"
+color red "\<(std(in|out|err)?|override|final|assert|errno)\>"
+
+# Operators and punctuators.
+color brightcyan "\<(defined|_Pragma|__has_include|_Generic)\>"
+color cyan "[].+*/%=@&|!^~?:,;(){}<>[-]"
+color cyan "\<((and|or|xor|not)(_eq)?|bit(and|or)|compl)\>"
+
+# Literals.
+color yellow "'([^'\]|(\\["'abefnrtv\\]))'" "'\\(([0-3]?[0-7]{1,2}))'" "'\\x[0-9A-Fa-f]{1,2}'"
+color yellow ""(\\.|[^\\"])*?""
+icolor yellow "\<[0-9]+(\.[0-9]*)?(e[+-]?[0-9]+)?(ul{0,2}|l{1,2}u?|f)?\>"
+icolor yellow "\<0x[0-9a-f]+(\.[0-9a-f]*)?(p[+-]?[0-9]+)?(ul{0,2}|l{1,2}u?|f)?\>"
+icolor yellow "\<0b[01]+(ul{0,2}|l{1,2}u?)?\>"
 
 # Preprocessor directives.
-color brightcyan start="^[[:blank:]]*#[[:blank:]]*((el)?if(n?def)?|error|pragma|warning)\>" end="(\`|[^\])$"
-color brightcyan "^[[:blank:]]*#[[:blank:]]*((define|else|embed|endif|include(_next)?|line|undef)\>|$)"
+color brightgreen "^[[:space:]]*#[[:space:]]*include(_next)?[[:space:]]+((<[^>]*>)|("[^"]*"))"
+color brightcyan start="^[[:space:]]*#[[:space:]]*(if(n?def)?|elif|define|undef|warning|error|pragma)\>" end="(\`|[^\\])$"
+color brightcyan "^[[:space:]]*#[[:space:]]*(else|endif|include(_next)?)?\>"
+
+# GCC builtins.
+###color cyan "__attribute__[[:space:]]*\(\([^)]*\)\)" "__(aligned|asm|builtin|hidden|inline|packed|restrict|section|typeof|weak)__"
 
 # Comments.
 color brightblue "//[^"]*$|(^|[[:blank:]])//.*"
 color brightblue start="/\*" end="\*/"
 
 # Reminders.
-color brightwhite,yellow "\<(FIXME|TODO|XXX)\>"
+color red,yellow "\<(FIXME|TODO|XXX)\>"
 
 # Trailing whitespace.
 color ,green "[[:space:]]+$"
